src/OpenTelemetry/Internal/CircularBuffer.cs

```diff
@@ -144,8 +147,10 @@ public bool TryAdd(T value, int maxSpinCount)
                     continue;
                 }
 
-                var index = (int)(head % this.Capacity);
-                this.trait[index] = value;
+                var previous = Interlocked.Exchange(ref this.trait[headSnapshot % this.Capacity], value);
```

Comment:
headSnapshot would be incorrect? we need this.trait[this.head]...
Not sure i missed something!

Reply:
If you look at the before (on the left) line 137 when we go to access the array we have already verified `head == headSnapshot`. I thought it was a bit confusing to have this.head (field), headSnapshot (local copy), & head (local verified alias to headSnapshot) so I simplified it a bit. Could be I made it more confusing 😄

Line 140 is the important part. If we fall into the spin it means some other thread took head and our local headSnapshot cannot be used. If we don't spin because we verified our headSnapshot was the true head, then this thread is safe to write to headSnapshot.